### ISSUE TITLE
Test if API and local SQL queries result in same objects

### ISSUE THREAD

**Issue by PietrH** [inbo/etn#330]
During #317 I noticed 9 functions return a different result between the API and a local SQL query. 

- [x] Add tests all exported functions to see if API and local queries match
- [x] Create custom expectation for comparing api to sql query
- [x] find why there is a mismatch, and fix it!

Some of these differences _might_ be fixed, or at least different after merging #328

## API - SQL mismatch

mismatches for:

- get_cpod_projects
- list_animal_project_codes
- list_acoustic_project_codes
- get_animals
- list_receiver_ids
- get_tags
- get_animal_projects
- list_scientific_names
- get_acoustic_projects

---

              Maybe this is a good use case for a [custom expectation](https://testthat.r-lib.org/articles/custom-expectation.html).

For example, you could create a `expect_call_agnostic()`, with one argument: a function call. The helper could grab the function call (and its arguments), run it locally and over API and expect the result to be the same.

It could then be called as part of a regular test-function.R file:

```R
testthat("get_animals() returns same result locally and via API", {
  expect_call_agnostic(get_animals(animal_id = "5"))
})

_Originally posted by @peterdesmet in https://github.com/inbo/etn/issues/317#issuecomment-2421841943_

**Comment by PietrH** [comment]
lil script to check for these functions:

```r
for(function_to_test in getNamespaceExports("etnservice")) {
  test_that(paste(
    function_to_test,
    "returns same result over api as over local db connection"
  ),
  {
    skip_if_offline()
    skip("not all functions work with no arguments")

    api_result <- do.call(function_to_test, list(api = TRUE))
    sql_result <- do.call(function_to_test, list(api = FALSE))
    expect_identical(api_result, sql_result, label = function_to_test)
  })
}

for(function_to_test in getNamespaceExports("etnservice")){
  print(function_to_test)
  api_result <- do.call(function_to_test, list(api = TRUE))
  sql_result <- do.call(function_to_test, list(api = FALSE))
  expect_identical(api_result, sql_result, label = function_to_test)
}

test_result <- purrr::map(
  getNamespaceExports("etnservice"),
  purrr::safely(function(function_to_test){
    api_result <- do.call(function_to_test, list(api = TRUE))
    sql_result <- do.call(function_to_test, list(api = FALSE))
    expect_identical(api_result, sql_result)
  }
)) %>% purrr::set_names(getNamespaceExports("etnservice"))

# only keep those with errors
test_result[purrr::map_lgl(test_result, ~!is.null(purrr::pluck(.x,"error")))]


```

**Comment by PietrH** [comment]
```r
skip_if(Sys.info()["nodename"] != "rshiny", "Test is not running on Lifewatch RStudio Server")
```

Or even better:

```r
skip_if("ETN" %in% odbc::odbcListDataSources()$name, "ETN is not a local database on this machine")
```

**Comment by PietrH** [comment]
Blocked by #423 , I can only run these tests when I have local database access.

**Comment by PietrH** [comment]
Mismatches should no longer occur, as local database queries use etnservice as a dependency. As long as the local etnservice version is identical to the one deployed on OpenCPU, the results should be identical. 

Only possible exceptions are: 

- `get_acoustic_detections()` uses internal logic
- `write_dwc()` `download_acoustic_dataset()` : client side processing

**Comment by PietrH** [comment]
As far as I can see this is now resolved, if any mismatches are found I can reopen or create new issues.

**Comment by PietrH** [comment]
Nope, #421 is the PR that will close this.

**Comment by PietrH** [comment]
This issue is about checking for differences, I should make a different Issue for fixing them, and split it into smaller tasks

**Comment by PietrH** [comment]
Two tests are failing because the public opencpu instance at [cloud.opencpu.org](url) returns a HTTP 526 due to Cloudfare SSL certificate trouble. 

The other tests fail because of actual differences between API and local database differences.
